### PR TITLE
feat(reporting): add BackfillBasicReportReportingSets tool

### PR DESCRIPTION
**Pull request by Marco-Premier**
# SUMMARY

BasicReports written via the internal InsertBasicReport method before external_reporting_set_id was validated have it unset on their ReportingUnitComponentSummary entries.

The tool sweeps every SUCCEEDED BasicReport across all MeasurementConsumers. It resolves each affected component summary to the Campaign Group child ReportingSet whose membership equals the component's event_group_summaries, and creates one for a membership that has no match. Pass --dry-run to report without writing.

# TEST

Stress-tested on dedicated Spanner and Cloud SQL instances against 20,000 BasicReports (15,000 broken, 5,000 already valid, result details up to 1.5 MiB): the apply run backfilled all 15,000 in 33m26s at 0.99 GiB peak resident memory with 0 skipped, 0 unresolved and 75 ReportingSets reused / 75 created exactly matching the seeded fixture, the preceding --dry-run predicted every counter identically, and a re-run afterwards reported alreadyValid=20000, updated=0, created=0, confirming the backfill is complete and idempotent.

Issue: #4289

### CODE REVIEW

**Comment by wfa-reviewable** [comment]
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352)

File: src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        componentSummary.externalReportingSetId = existingId
+        componentSummariesFilled++
+        // A ReportingSet this run created and then found again in the index is not a reuse.
+        if (existingId !in mintedReportingSetIds) {
+          reusedReportingSetIds.add(existingId)
+        }
+        continue
+      }
+
+      val mintedId: String = mintReportingSet(campaignGroupKey, membership)
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt:306]: This mints a ReportingSet in Postgres before the loop has finished checking whether every component summary in this BasicReport is resolvable. If a later component summary in the same report turns out unresolved (empty event_group_summaries), the unresolved > 0 check below (line 289) skips the Spanner write entirely, but the ReportingSet just created here for an earlier component stays committed in Postgres, permanently orphaned and referenced by nothing.

This contradicts the docstring's and README's stated guarantee that a BasicReport is written only when all of its component summaries resolve — that's true for the Spanner write but not for this Postgres write. Real BasicReports commonly have multiple ReportingUnitComponentSummary entries (one per DataProvider), so a mix of resolvable and unresolvable components in one report is not a rare case. It's also untested: every case in BasicReportReportingSetBackfillerTest.kt uses insertTestBasicReport(), which hardcodes exactly one component summary per report.

Split backfillBasicReport into two passes: first resolve every component summary's membership and look up existing matches without minting anything; only if none are unresolved, do a second pass that mints any still-missing ReportingSets and applies the IDs. That keeps the Postgres side effects gated on the same all-or-nothing guarantee as the Spanner write, and add a test with two component summaries in one BasicReport — one resolvable-but-new, one with empty event_group_summaries — asserting reportingSetsCreated == 0 after the run.

**Review by stevenwarejones — COMMENTED** [review]
Non-atomic Postgres side effect when a BasicReport ends up skipped.

File: src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+ * Every SUCCEEDED `BasicReport`, across all `MeasurementConsumer`s, is examined. Each component
+ * summary missing the field is resolved to the Campaign Group child `ReportingSet` whose membership
+ * equals the component's `event_group_summaries`; a `ReportingSet` is created for a membership that
+ * has no match.
+ *
+ * A `BasicReport` is written only when all of its component summaries resolve.
+ *
+ * @param dryRun when true, no write is issued to either database
+ */
+class BasicReportReportingSetBackfiller(
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt:66]: nit: Nothing prevents two concurrent invocations of this tool from racing on the same Campaign Group. Each run builds its own in-memory reportingSetIdsByCampaignGroup cache from readCampaignGroupMembershipIndex; if two instances both read that index before either has minted a ReportingSet for the same missing membership, both will call CreateReportingSet and end up with two ReportingSets carrying identical membership under the same Campaign Group. This will not throw (the minted external_reporting_set_id is a random UUID, so there is no unique-constraint collision) — it just silently defeats the reuse invariant the rest of the tool relies on, and neither run's summary would report anything wrong.

Since this is a manually-invoked operator tool with no lock/lease mechanism, call this out explicitly in the README (e.g. do not run two instances against the same environment concurrently) so an operator does not learn about it by finding duplicate ReportingSets later.

**Review by stevenwarejones — COMMENTED** [review]
No guard against concurrent invocations of the tool.

File: src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+  private fun eventGroupKey(
+    cmmsEventGroupId: String,
+    cmmsDataProviderId: String = CMMS_DATA_PROVIDER_ID,
+  ) =
+    ReportingSetKt.PrimitiveKt.eventGroupKey {
+      this.cmmsDataProviderId = cmmsDataProviderId
+      this.cmmsEventGroupId = cmmsEventGroupId
+    }
+
+  private suspend fun insertTestBasicReport(
```
Comment (stevenwarejones) [inline src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt:315]: insertTestBasicReport hardcodes exactly one ReportingUnitComponentSummary per BasicReport, and readComponentSummaryReportingSetId() below calls .single() on the component summary list. Together these mean the test suite cannot express a BasicReport with more than one component summary at all.

Real BasicReports are cross-media reports over multiple DataProviders, so multiple component summaries per report is the common case, not the exception. Its the exact shape needed to catch the mint-before-full-resolution-check bug flagged elsewhere in this review (BasicReportReportingSetBackfiller.kt:280): a blank-but-resolvable component mints a real ReportingSet before a second blank-but-unresolvable component in the same report is found, so the whole report ends up skipped with an orphaned ReportingSet left behind.

Extend insertTestBasicReport to accept a list of component summaries (or add a second helper) and add two tests: (1) one BasicReport with two component summaries, one already-valid and one blank-but-new, asserting the already-valid one keeps its exact externalReportingSetId untouched; (2) one BasicReport mixing a resolvable-blank and an unresolvable-blank component, asserting reportingSetsCreated == 0. Neither case can be expressed with the current helper.

**Review by stevenwarejones — COMMENTED** [review]
No test coverage for a BasicReport with multiple component summaries.

File: src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+import org.wfanet.measurement.reporting.deploy.v2.gcloud.spanner.db.insertBasicReport
+import org.wfanet.measurement.reporting.deploy.v2.gcloud.spanner.db.insertMeasurementConsumer
+import org.wfanet.measurement.reporting.deploy.v2.gcloud.spanner.testing.Schemata
+import org.wfanet.measurement.reporting.deploy.v2.postgres.PostgresMeasurementConsumersService
+import org.wfanet.measurement.reporting.deploy.v2.postgres.PostgresReportingSetsService
+import org.wfanet.measurement.reporting.deploy.v2.postgres.readers.ReportingSetReader
+import org.wfanet.measurement.reporting.deploy.v2.postgres.testing.Schemata as PostgresSchemata
+
+@RunWith(JUnit4::class)
+class BasicReportReportingSetBackfillerTest {
```
Comment (stevenwarejones) [inline src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt:65]: Nothing in this suite, or anywhere else touched by this PR, exercises the actual read path -- SpannerBasicReportsService or BasicReportProtoConversions.toBasicReport() -- against a report this tool has backfilled. That matters more than it looks.

e8c347ba6 (#4336) already made that read path tolerate a blank external_reporting_set_id by silently omitting reporting_set instead of throwing, with a TODO to remove that workaround once the backfill in this PR lands. That means the original crash this tool exists to fix is already gone, independent of whether this backfill works correctly. If the backfill silently fails to update some reports, or writes a wrong value, nothing will crash -- GetBasicReport/ListBasicReports will just keep quietly omitting reporting_set (or return a wrong one) with no error signal anywhere. This tools own unit tests, which only assert against raw Spanner/Postgres reads, are the only thing standing between a subtle bug here and it going unnoticed in production indefinitely.

Add a test that runs the backfiller against a report with a blank component summary, then reads the result back through SpannerBasicReportsService.getBasicReport (or BasicReportProtoConversions.toBasicReport() directly) and asserts reporting_set is now populated with the correct resource name. Thats the only thing that would actually prove issue #4289 is resolved from the callers point of view, rather than just proving the Spanner row changed.

**Review by stevenwarejones — COMMENTED** [review]
No test confirms the fix through the actual read path.

**Review by stevenwarejones — COMMENTED** [review]
> @stevenwarejones reviewed 12 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352)* status: 12 of 17 files reviewed, 4 unresolved discussions (waiting on kungfucraig and Marco-Premier).

**Review by stevenwarejones — COMMENTED** [review]
Please add both these tests

> @stevenwarejones made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352#-:-P-Qlm1TCKVXQPWXpTcC:b-qlhedn)* status: 12 of 17 files reviewed, 4 unresolved discussions (waiting on kungfucraig and Marco-Premier).

File: src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt (hunk 1 above)
Comment (Marco-Premier) [reply src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt:306]: Done.

File: src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt (hunk 4 above)
Comment (Marco-Premier) [reply src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt:65]: Done.

File: src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt (hunk 3 above)
Comment (Marco-Premier) [reply src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt:315]: Done.

**Review by Marco-Premier — COMMENTED** [review]
> @Marco-Premier made 3 comments and resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352)* status: 12 of 17 files reviewed, 3 unresolved discussions (waiting on kungfucraig and stevenwarejones).

File: src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt (hunk 2 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt:66]: Verified — documented in both the class KDoc and a new README Concurrency section: do not run two instances against the same environment concurrently, since there is no lock and duplicate ReportingSets with identical membership can result without either run reporting an error. That is the right level of fix for a manually-invoked ops tool.

File: src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt (hunk 1 above)
Comment (stevenwarejones) [reply src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfiller.kt:306]: Verified — resolution is now a separate pass from minting (memberships computed and the unresolved check run to completion before any mintReportingSet call), so a report with a resolvable and an unresolvable component summary no longer leaves an orphaned ReportingSet behind. Confirmed by reading the new two-pass structure and the new test asserting reportingSetsCreated == 0 with the campaign group ReportingSets list containing only the campaign group itself.

File: src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt (hunk 4 above)
Comment (stevenwarejones) [reply src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt:65]: Verified — the new test reads a backfilled BasicReport back through InternalBasicReport.toBasicReport(), the actual v2alpha conversion the public read path uses, and asserts reporting_set is populated with the correct resource name. That directly proves the fix from the callers point of view rather than only proving the Spanner row changed.

File: src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt (hunk 3 above)
Comment (stevenwarejones) [reply src/test/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/BasicReportReportingSetBackfillerTest.kt:315]: Verified — insertTestBasicReport now takes a List<ComponentSpec> and can seed multiple component summaries per BasicReport. The new tests cover exactly the two scenarios asked for: one mixing a resolvable and an unresolvable component (asserting no orphaned ReportingSet), and one mixing an already-valid and a blank component (asserting the already-valid one keeps its exact external_reporting_set_id while its sibling is backfilled).

**Review by stevenwarejones — CHANGES_REQUESTED** [review]
Verified the test coverage added in b1016c218: the multi-component-summary fixtures cover both the mint-before-resolution-check bug and the mixed already-valid/blank case, and the new test reading a backfilled report back through toBasicReport() confirms the fix through the actual read path. All four review threads verified and marked satisfied.

**Review by stevenwarejones — APPROVED** [review]
:lgtm:

> @stevenwarejones reviewed 3 files and all commit messages, made 1 comment, and resolved 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352#-:-P-UIdTrA6kM7LgCN_Hk:bnfp4nl)* status: 15 of 17 files reviewed, 1 unresolved discussion (waiting on kungfucraig).

File: src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/README.md [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+Backfills
+`ResultGroup.MetricMetadata.ReportingUnitComponentSummary.external_reporting_set_id`
+on stored `BasicReport`s.
+
+`BasicReport`s written through the internal `InsertBasicReport` method before
+that field was validated have it unset. The public read path assembles a
+`ReportingSet` resource name from it, so such a `BasicReport` cannot be served
+by `GetBasicReport`, and fails the whole page for `ListBasicReports`. See
+[issue #4289](https://github.com/world-federation-of-advertisers/cross-media-measurement/issues/4289).
```
Comment (stevenwarejones) [inline src/main/kotlin/org/wfanet/measurement/reporting/deploy/v2/gcloud/spanner/tools/README.md:16]: BasicReportProtoConversions.kt:850 has a TODO citing this issue: remove the tolerate-blank check once BasicReports written via InsertBasicReport without external_reporting_set_id have been backfilled. This PR performs that backfill, but nothing will signal whether the condition recurs afterward. The write vector that produced the original corruption (an external caller, per #4289) has not been confirmed to go through InsertBasicReport itself -- if it instead writes to Spanner directly, #4306's RPC-level validation never touched it, and the tolerate-blank check is the only thing standing between a recurrence and a crash.

Add a logger.warning at the omission site (BasicReportProtoConversions.kt:850) so any post-backfill occurrence is observable instead of silent. Use that as the actual gate for closing the TODO -- only remove the tolerance once the warning has stayed silent for a sustained period, confirming the write vector is actually closed, rather than removing it on a timeline alone.

**Review by stevenwarejones — COMMENTED** [review]
Follow-up: gate removal of the tolerate-blank TODO on observability, not just this backfill.

**Review by stevenwarejones — COMMENTED** [review]
> @stevenwarejones reviewed 2 files and all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352)* status: all files reviewed, 2 unresolved discussions (waiting on kungfucraig and Marco-Premier).

**Review by stevenwarejones — COMMENTED** [review]
:lgtm:

> @stevenwarejones made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352#-:-P-UX5hPD7Fry59xsO90:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on kungfucraig and Marco-Premier).

**Review by stevenwarejones — APPROVED** [review]
> @stevenwarejones resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352)* status: all files reviewed, 1 unresolved discussion (waiting on kungfucraig and Marco-Premier).

**Review by kungfucraig — APPROVED** [review]
Confirming this does not meet the criteria for "critical change."

> @kungfucraig reviewed 18 files and all commit messages, and made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352#-:-P-V0zJ_B5pxEVKEy1n5:bvliun2)* status: all files reviewed, 1 unresolved discussion (waiting on Marco-Premier).

**Review by kungfucraig — APPROVED** [review]
:lgtm:

> @kungfucraig made 1 comment.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352#-:-P-V1EsLAlyS3v4-zExV:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on Marco-Premier).

**Review by Marco-Premier — COMMENTED** [review]
> @Marco-Premier resolved 1 discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on Marco-Premier).

**Review by Marco-Premier — COMMENTED** [review]
> @Marco-Premier partially reviewed 19 files.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/4352)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on kungfucraig, stevenwarejones, and stevenwarejones+AGNT).
